File: gpt_validator.py
"""
GPT-based PII Validation System
Uses Azure OpenAI GPT to validate whether detected PII is actually sensitive
"""

import requests
import os
import json
import re
from typing import Dict, List, Tuple, Optional
from dataclasses import dataclass
from dotenv import load_dotenv
import time

# Load environment variables
load_dotenv()

@dataclass
class ValidationResult:
    """Result of GPT validation"""
    is_real_pii: bool
    confidence: float
    explanation: str
    should_redact: bool
    suggested_action: str

class GPTPIIValidator:
    """
    Uses GPT-4o-mini to validate PII detection results
    Acts as a smart filter to reduce false positives
    """
    
    def __init__(self, azure_api_key: Optional[str] = None, azure_endpoint: Optional[str] = None, 
                 deployment_name: Optional[str] = None, api_version: Optional[str] = None):
        """
        Initialize GPT PII Validator with Azure OpenAI
        
        Args:
            azure_api_key: Azure API key (if None, loads from AZURE_KEY env var)
            azure_endpoint: Azure OpenAI endpoint (if None, loads from AZURE_GPT_ENDPOINT env var)
            deployment_name: Deployment name (if None, loads from AZURE_GPT_DEPLOYMENT env var)
            api_version: API version (if None, loads from AZURE_GPT_API_VERSION env var)
        """
        self.azure_api_key = azure_api_key or os.getenv('AZURE_KEY')
        self.azure_endpoint = azure_endpoint or os.getenv('AZURE_GPT_ENDPOINT', 'https://dgrif-malg2o2i-eastus2.cognitiveservices.azure.com')
        self.deployment_name = deployment_name or os.getenv('AZURE_GPT_DEPLOYMENT', 'gpt-5-chat')
        self.api_version = api_version or os.getenv('AZURE_GPT_API_VERSION', '2025-01-01-preview')
        
        if not self.azure_api_key:
            raise ValueError("Azure API key not found. Set AZURE_KEY environment variable or pass azure_api_key parameter.")
        
        self.base_url = f"{self.azure_endpoint}/openai/deployments/{self.deployment_name}/chat/completions"
        self.model = self.deployment_name
        self.cost_per_1k_tokens = 0.00015  # Approximate cost
        self.total_cost = 0.0
        
        # Validation prompts for different contexts
        self.base_prompt = """
You are a PII validation expert. Your job is to determine if text flagged as PII is actually sensitive personal information or just business terminology.

Context: This text was flagged by Azure AI as containing PII, but it might be a false positive.

Guidelines:
- REAL PII: Actual names, emails, phone numbers, addresses, SSNs, etc. of real people
- NOT PII: Job titles, user roles, business terms, product names, generic references
- BUSINESS TERMS: "users", "customers", "contacts", "agents", "admins" are usually NOT PII
- CONTEXT MATTERS: "external users" in support context is NOT PII

Examples:
- "John Smith called" → REAL PII (actual person name)
- "external users can't access" → NOT PII (user role/type)
- "contact form submission" → NOT PII (business process)
- "support agent assigned" → NOT PII (job role)
- "customer service team" → NOT PII (department)

Respond with JSON only:
{
    "is_real_pii": boolean,
    "confidence": 0.0-1.0,
    "explanation": "brief reason",
    "should_redact": boolean,
    "suggested_action": "keep_original|redact|partial_redact"
}
"""
    
    def validate_pii_detection(self, 
                              original_text: str, 
                              detected_entities: List[Dict],
                              context: str = "general") -> Dict[str, ValidationResult]:
        """
        Validate multiple PII detections in text using GPT
        
        Args:
            original_text: Original text before redaction
            detected_entities: List of detected PII entities from Azure
            context: Context hint (e.g., "support_ticket", "customer_data")
            
        Returns:
            Dictionary mapping entity text to ValidationResult
        """
        if not detected_entities:
            return {}
        
        results = {}
        
        # Group entities by type for batch processing
        entity_groups = {}
        for entity in detected_entities:
            entity_type = entity.get('category', 'Unknown')
            if entity_type not in entity_groups:
                entity_groups[entity_type] = []
            entity_groups[entity_type].append(entity)
        
        # Process each entity type
        for entity_type, entities in entity_groups.items():
            try:
                batch_results = self._validate_entity_batch(
                    original_text, entities, entity_type, context
                )
                results.update(batch_results)
            except Exception as e:
                print(f"Error validating {entity_type} entities: {e}")
                # Default to keeping redaction on error
                for entity in entities:
                    results[entity['text']] = ValidationResult(
                        is_real_pii=True,
                        confidence=0.5,
                        explanation=f"Validation failed: {str(e)}",
                        should_redact=True,
                        suggested_action="redact"
                    )
        
        return results
    
    def _validate_entity_batch(self, 
                              original_text: str,
                              entities: List[Dict],
                              entity_type: str,
                              context: str) -> Dict[str, ValidationResult]:
        """Validate a batch of entities of the same type"""
        
        # Create context-aware prompt
        context_prompt = self._get_context_prompt(context, entity_type)
        
        # Prepare entities for validation
        entity_texts = [entity['text'] for entity in entities]
        
        validation_prompt = f"""
{self.base_prompt}

{context_prompt}

Original text: "{original_text}"

Entity type flagged: {entity_type}
Entities to validate: {entity_texts}

For each entity, determine if it's real PII or a false positive.
"""
        
        try:
            print(f"🤖 Making Azure GPT API call to {self.base_url}")
            print(f"🤖 Validating {len(entity_texts)} entities of type {entity_type}")
            
            # Prepare Azure OpenAI request
            headers = {
                "Content-Type": "application/json",
                "Authorization": f"Bearer {self.azure_api_key}"
            }
            
            payload = {
                "messages": [
                    {"role": "system", "content": "You are a PII validation expert."},
                    {"role": "user", "content": validation_prompt}
                ],
                "max_tokens": 500,
                "temperature": 0.1  # Low temperature for consistent results
            }
            
            # Call Azure GPT
            response = requests.post(
                f"{self.base_url}?api-version={self.api_version}",
                headers=headers,
                json=payload,
                timeout=30
            )
            
            print(f"🤖 Azure GPT response status: {response.status_code}")
            
            if response.status_code != 200:
                raise Exception(f"Azure OpenAI request failed: {response.status_code} - {response.text}")
            
            response_data = response.json()
            
            # Calculate cost (approximate)
            tokens_used = response_data.get('usage', {}).get('total_tokens', 100)  # Fallback estimate
            cost = (tokens_used / 1000) * self.cost_per_1k_tokens
            self.total_cost += cost
            
            print(f"🤖 GPT validation cost: ${cost:.6f} (Total: ${self.total_cost:.6f})")
            
            # Parse response
            response_text = response_data['choices'][0]['message']['content']
            return self._parse_validation_response(response_text, entities)
            
        except Exception as e:
            print(f"GPT validation error: {e}")
            # Default to keeping redaction
            return {
                entity['text']: ValidationResult(
                    is_real_pii=True,
                    confidence=0.5,
                    explanation=f"GPT validation failed: {str(e)}",
                    should_redact=True,
                    suggested_action="redact"
                ) for entity in entities
            }
    
    def _get_context_prompt(self, context: str, entity_type: str) -> str:
        """Get context-specific validation prompt"""
        
        context_prompts = {
            "support_ticket": f"""
CONTEXT: Support ticket/customer service data
- "users", "customers", "contacts" are usually user TYPES, not individual people
- "external users", "internal users", "admin users" are ROLES, not PII
- "agent", "support staff", "customer service" are JOB TITLES, not PII
- Only actual individual names like "John Smith" should be considered PII
- System/role references should NOT be redacted

Entity type: {entity_type}
Special attention: PersonType entities are often false positives in support context.
""",
            
            "zendesk": f"""
CONTEXT: Zendesk support ticket data
- Very high likelihood of false positives for PersonType
- "Co-managed users", "external users", "end users" are USER CATEGORIES, not PII
- "contact", "customer", "agent" in business context are ROLES, not individual people
- "user interface", "user experience" are PRODUCT TERMS, not PII
- Only redact if it's clearly an individual person's name

Entity type: {entity_type}
""",
            
            "customer_data": f"""
CONTEXT: Customer database/CRM data
- Higher likelihood that Person entities are real individual names
- But still check for generic terms like "customer", "client", "user"
- Email domains and system emails may not be PII

Entity type: {entity_type}
""",
            
            "general": f"""
CONTEXT: General business data
- Consider business vs personal context
- Job titles, roles, and categories are typically not PII

Entity type: {entity_type}
"""
        }
        
        return context_prompts.get(context, context_prompts["general"])
    
    def _parse_validation_response(self, response_text: str, entities: List[Dict]) -> Dict[str, ValidationResult]:
        """Parse GPT response and create ValidationResult objects"""
        
        results = {}
        
        try:
            # Try to parse as JSON
            if response_text.strip().startswith('{'):
                # Single entity response
                parsed = json.loads(response_text)
                if entities:
                    entity_text = entities[0]['text']
                    results[entity_text] = ValidationResult(
                        is_real_pii=parsed.get('is_real_pii', True),
                        confidence=parsed.get('confidence', 0.5),
                        explanation=parsed.get('explanation', 'No explanation'),
                        should_redact=parsed.get('should_redact', True),
                        suggested_action=parsed.get('suggested_action', 'redact')
                    )
            
            elif response_text.strip().startswith('['):
                # Multiple entities response
                parsed_list = json.loads(response_text)
                for i, parsed in enumerate(parsed_list):
                    if i < len(entities):
                        entity_text = entities[i]['text']
                        results[entity_text] = ValidationResult(
                            is_real_pii=parsed.get('is_real_pii', True),
                            confidence=parsed.get('confidence', 0.5),
                            explanation=parsed.get('explanation', 'No explanation'),
                            should_redact=parsed.get('should_redact', True),
                            suggested_action=parsed.get('suggested_action', 'redact')
                        )
            
            else:
                # Try to extract JSON from text
                json_match = re.search(r'\{.*\}', response_text, re.DOTALL)
                if json_match:
                    parsed = json.loads(json_match.group())
                    if entities:
                        entity_text = entities[0]['text']
                        results[entity_text] = ValidationResult(
                            is_real_pii=parsed.get('is_real_pii', True),
                            confidence=parsed.get('confidence', 0.5),
                            explanation=parsed.get('explanation', 'No explanation'),
                            should_redact=parsed.get('should_redact', True),
                            suggested_action=parsed.get('suggested_action', 'redact')
                        )
        
        except json.JSONDecodeError:
            # Fallback: parse as natural language
            for entity in entities:
                entity_text = entity['text']
                
                # Simple keyword analysis of GPT response
                response_lower = response_text.lower()
                entity_lower = entity_text.lower()
                
                # Look for clear indicators
                is_not_pii = any(phrase in response_lower for phrase in [
                    f"{entity_lower} is not pii",
                    f"{entity_lower} is not personal",
                    "false positive",
                    "not sensitive",
                    "business term",
                    "job title",
                    "role",
                    "category"
                ])
                
                is_pii = any(phrase in response_lower for phrase in [
                    f"{entity_lower} is pii",
                    f"{entity_lower} is personal",
                    "should be redacted",
                    "sensitive",
                    "individual name"
                ])
                
                if is_not_pii:
                    should_redact = False
                    confidence = 0.8
                elif is_pii:
                    should_redact = True
                    confidence = 0.8
                else:
                    should_redact = True  # Default to safe
                    confidence = 0.3
                
                results[entity_text] = ValidationResult(
                    is_real_pii=should_redact,
                    confidence=confidence,
                    explanation="Parsed from natural language response",
                    should_redact=should_redact,
                    suggested_action="redact" if should_redact else "keep_original"
                )
        
        # Ensure all entities have results
        for entity in entities:
            entity_text = entity['text']
            if entity_text not in results:
                results[entity_text] = ValidationResult(
                    is_real_pii=True,
                    confidence=0.5,
                    explanation="Could not parse GPT response",
                    should_redact=True,
                    suggested_action="redact"
                )
        
        return results
    
    def apply_validation_results(self, 
                                original_text: str,
                                detected_entities: List[Dict],
                                validation_results: Dict[str, ValidationResult]) -> Tuple[str, List[Dict]]:
        """
        Apply validation results to create a corrected redaction
        
        Args:
            original_text: Original text
            detected_entities: Original detected entities
            validation_results: GPT validation results
            
        Returns:
            Tuple of (corrected_text, filtered_entities)
        """
        
        # Filter entities based on validation
        valid_entities = []
        for entity in detected_entities:
            entity_text = entity['text']
            validation = validation_results.get(entity_text)
            
            if validation and validation.should_redact:
                valid_entities.append(entity)
            # Entities that shouldn't be redacted are filtered out
        
        # Rebuild redacted text with only validated entities
        corrected_text = original_text
        
        # Sort entities by position (reverse order to maintain positions)
        sorted_entities = sorted(valid_entities, key=lambda e: e['offset'], reverse=True)
        
        for entity in sorted_entities:
            start = entity['offset']
            end = start + entity['length']
            
            # Use original redaction label
            redaction_label = entity.get('redaction', f"[REDACTED_{entity['category'].upper()}]")
            
            corrected_text = corrected_text[:start] + redaction_label + corrected_text[end:]
        
        return corrected_text, valid_entities
    
    def get_cost_summary(self) -> Dict[str, float]:
        """Get cost summary for GPT validation"""
        return {
            'total_cost': self.total_cost,
            'cost_per_validation': self.cost_per_1k_tokens,
            'model': self.model
        }

# Quick test function
def test_validator():
    """Test the GPT validator with sample data"""
    
    # Sample problematic text
    test_text = "Co-managed users can not see note that comes in internal only if the external user is not a contact"
    
    # Sample detected entities (what Azure would return)
    test_entities = [
        {
            'text': 'users',
            'category': 'PersonType',
            'confidence': 0.8,
            'offset': 11,
            'length': 5,
            'redaction': '[REDACTED_PERSONTYPE]'
        },
        {
            'text': 'user',
            'category': 'PersonType', 
            'confidence': 0.7,
            'offset': 81,
            'length': 4,
            'redaction': '[REDACTED_PERSONTYPE]'
        },
        {
            'text': 'contact',
            'category': 'PersonType',
            'confidence': 0.6,
            'offset': 98,
            'length': 7,
            'redaction': '[REDACTED_PERSONTYPE]'
        }
    ]
    
    try:
        validator = GPTPIIValidator()
        
        print("Testing GPT PII Validator...")
        print(f"Original: {test_text}")
        print(f"Detected entities: {[e['text'] for e in test_entities]}")
        
        # Validate
        validation_results = validator.validate_pii_detection(
            test_text, test_entities, context="support_ticket"
        )
        
        print("\nValidation Results:")
        for entity_text, result in validation_results.items():
            print(f"  '{entity_text}': {'❌ NOT PII' if not result.should_redact else '✅ IS PII'} "
                  f"(confidence: {result.confidence:.2f}) - {result.explanation}")
        
        # Apply corrections
        corrected_text, filtered_entities = validator.apply_validation_results(
            test_text, test_entities, validation_results
        )
        
        print(f"\nCorrected: {corrected_text}")
        print(f"Cost: ${validator.get_cost_summary()['total_cost']:.4f}")
        
        return True
        
    except Exception as e:
        print(f"Test failed: {e}")
        return False

if __name__ == "__main__":
    test_validator()

File: redactify/backend/internal/pii/gpt_validator.go
package pii

import (
	"bytes"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"redactify/pkg/config"
	"strings"
	"time"
)

// GPTValidator handles GPT-based PII validation
type GPTValidator struct {
	config *config.Config
	client *http.Client
}

// ValidationResult represents the result of GPT validation
type ValidationResult struct {
	IsRealPII       bool    `json:"is_real_pii"`
	Confidence      float64 `json:"confidence"`
	Explanation     string  `json:"explanation"`
	ShouldRedact    bool    `json:"should_redact"`
	SuggestedAction string  `json:"suggested_action"`
}

// GPTRequest represents the request to Azure OpenAI
type GPTRequest struct {
	Messages    []GPTMessage `json:"messages"`
	MaxTokens   int          `json:"max_tokens"`
	Temperature float64      `json:"temperature"`
	TopP        float64      `json:"top_p"`
}

// GPTMessage represents a message in the conversation
type GPTMessage struct {
	Role    string `json:"role"`
	Content string `json:"content"`
}

// GPTResponse represents the response from Azure OpenAI
type GPTResponse struct {
	Choices []struct {
		Message struct {
			Content string `json:"content"`
		} `json:"message"`
	} `json:"choices"`
	Error *struct {
		Message string `json:"message"`
		Type    string `json:"type"`
	} `json:"error,omitempty"`
}

// NewGPTValidator creates a new GPT validator
func NewGPTValidator(cfg *config.Config) *GPTValidator {
	return &GPTValidator{
		config: cfg,
		client: &http.Client{
			Timeout: 30 * time.Second,
		},
	}
}

// ValidateEntities validates a list of detected entities using GPT
func (v *GPTValidator) ValidateEntities(text string, entities []Entity) ([]ValidationResult, error) {
	if v.config.Azure.GPTEndpoint == "" || v.config.Azure.GPTAPIKey == "" {
		// Return default validation if GPT not configured
		results := make([]ValidationResult, len(entities))
		for i := range entities {
			results[i] = ValidationResult{
				IsRealPII:       true,
				Confidence:      0.8, // Default confidence
				Explanation:     "GPT validation not configured, using default approval",
				ShouldRedact:    true,
				SuggestedAction: "redact",
			}
		}
		return results, nil
	}

	var results []ValidationResult
	
	// Process entities in batches to avoid token limits
	batchSize := 5
	for i := 0; i < len(entities); i += batchSize {
		end := i + batchSize
		if end > len(entities) {
			end = len(entities)
		}
		
		batch := entities[i:end]
		batchResults, err := v.validateBatch(text, batch)
		if err != nil {
			// If GPT fails, default to allowing redaction
			for range batch {
				results = append(results, ValidationResult{
					IsRealPII:       true,
					Confidence:      0.7,
					Explanation:     fmt.Sprintf("GPT validation failed: %v", err),
					ShouldRedact:    true,
					SuggestedAction: "redact",
				})
			}
		} else {
			results = append(results, batchResults...)
		}
	}
	
	return results, nil
}

// validateBatch validates a batch of entities
func (v *GPTValidator) validateBatch(text string, entities []Entity) ([]ValidationResult, error) {
	// Create the validation prompt
	prompt := v.createValidationPrompt(text, entities)
	
	// Prepare GPT request
	request := GPTRequest{
		Messages: []GPTMessage{
			{
				Role:    "system",
				Content: v.getSystemPrompt(),
			},
			{
				Role:    "user",
				Content: prompt,
			},
		},
		MaxTokens:   1000,
		Temperature: 0.1, // Low temperature for consistent results
		TopP:        0.1,
	}
	
	// Make the API call
	response, err := v.callGPTAPI(request)
	if err != nil {
		return nil, err
	}
	
	// Parse the response
	return v.parseValidationResponse(response, len(entities))
}

// createValidationPrompt creates a prompt for validating the entities
func (v *GPTValidator) createValidationPrompt(text string, entities []Entity) string {
	var prompt strings.Builder
	
	prompt.WriteString("Please analyze the following text and validate whether the detected entities are actually sensitive PII that should be redacted:\n\n")
	prompt.WriteString("**Original Text:**\n")
	prompt.WriteString(text)
	prompt.WriteString("\n\n**Detected Entities:**\n")
	
	for i, entity := range entities {
		prompt.WriteString(fmt.Sprintf("%d. \"%s\" (Type: %s, Confidence: %.2f)\n", 
			i+1, entity.Text, entity.Type, entity.Confidence))
	}
	
	prompt.WriteString("\n**Instructions:**\n")
	prompt.WriteString("For each entity, respond with a JSON object containing:\n")
	prompt.WriteString("- is_real_pii: true/false\n")
	prompt.WriteString("- confidence: 0.0-1.0\n")
	prompt.WriteString("- explanation: brief reason\n")
	prompt.WriteString("- should_redact: true/false\n")
	prompt.WriteString("- suggested_action: \"redact\", \"keep\", or \"partial\"\n\n")
	prompt.WriteString("Return only a JSON array with one object per entity, in order.")
	
	return prompt.String()
}

// getSystemPrompt returns the system prompt for GPT
func (v *GPTValidator) getSystemPrompt() string {
	return `You are an expert PII (Personally Identifiable Information) validation system. Your job is to analyze detected entities and determine if they are actually sensitive PII that should be redacted.

Consider these factors:
1. Context: Is this information used in a business/professional context?
2. Sensitivity: Would this information pose a privacy risk if disclosed?
3. False Positives: Common business terms that look like PII but aren't sensitive
4. Placeholder Text: Example data that isn't real PII

Examples of what to KEEP (not redact):
- Business contact names in signatures
- Company names that happen to sound like person names
- Generic examples like "John Doe" or "example@company.com"
- Job titles and department names
- Product names or services

Examples of what to REDACT:
- Customer personal information
- Employee personal details in non-business context
- Real email addresses of individuals
- Phone numbers for personal contacts
- Addresses for individuals

Respond only with valid JSON arrays. Be conservative - when in doubt, choose to redact.`
}

// callGPTAPI makes the actual API call to Azure OpenAI
func (v *GPTValidator) callGPTAPI(request GPTRequest) (string, error) {
	// Build the URL
	url := fmt.Sprintf("%s/openai/deployments/%s/chat/completions?api-version=%s",
		strings.TrimSuffix(v.config.Azure.GPTEndpoint, "/"),
		v.config.Azure.GPTDeployment,
		v.config.Azure.GPTAPIVersion)
	
	// Marshal request to JSON
	jsonData, err := json.Marshal(request)
	if err != nil {
		return "", fmt.Errorf("failed to marshal request: %w", err)
	}
	
	// Create HTTP request
	httpReq, err := http.NewRequest("POST", url, bytes.NewBuffer(jsonData))
	if err != nil {
		return "", fmt.Errorf("failed to create request: %w", err)
	}
	
	// Set headers
	httpReq.Header.Set("Content-Type", "application/json")
	httpReq.Header.Set("api-key", v.config.Azure.GPTAPIKey)
	
	// Execute request
	resp, err := v.client.Do(httpReq)
	if err != nil {
		return "", fmt.Errorf("request failed: %w", err)
	}
	defer resp.Body.Close()
	
	// Read response
	body, err := io.ReadAll(resp.Body)
	if err != nil {
		return "", fmt.Errorf("failed to read response: %w", err)
	}
	
	// Check status code
	if resp.StatusCode != http.StatusOK {
		return "", fmt.Errorf("API returned status %d: %s", resp.StatusCode, string(body))
	}
	
	// Parse response
	var gptResp GPTResponse
	if err := json.Unmarshal(body, &gptResp); err != nil {
		return "", fmt.Errorf("failed to parse response: %w", err)
	}
	
	// Check for errors
	if gptResp.Error != nil {
		return "", fmt.Errorf("GPT API error: %s", gptResp.Error.Message)
	}
	
	// Return content
	if len(gptResp.Choices) > 0 {
		return gptResp.Choices[0].Message.Content, nil
	}
	
	return "", fmt.Errorf("no response from GPT")
}

// parseValidationResponse parses the GPT response into ValidationResult objects
func (v *GPTValidator) parseValidationResponse(response string, expectedCount int) ([]ValidationResult, error) {
	// Clean up the response (remove markdown code blocks if present)
	response = strings.TrimSpace(response)
	response = strings.TrimPrefix(response, "```json")
	response = strings.TrimPrefix(response, "```")
	response = strings.TrimSuffix(response, "```")
	response = strings.TrimSpace(response)
	
	var results []ValidationResult
	if err := json.Unmarshal([]byte(response), &results); err != nil {
		return nil, fmt.Errorf("failed to parse validation response: %w", err)
	}
	
	// Ensure we have the expected number of results
	if len(results) != expectedCount {
		return nil, fmt.Errorf("expected %d validation results, got %d", expectedCount, len(results))
	}
	
	return results, nil
}

// IsConfigured returns true if GPT validation is properly configured
func (v *GPTValidator) IsConfigured() bool {
	return v.config.Azure.GPTEndpoint != "" && v.config.Azure.GPTAPIKey != ""
}